File: android/src/main/java/com/logicwind/inappupdate/RnInAppUpdateModule.java
package com.logicwind.inappupdate;

import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;
import android.util.Log;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.google.android.play.core.appupdate.AppUpdateInfo;
import com.google.android.play.core.appupdate.AppUpdateManager;
import com.google.android.play.core.appupdate.AppUpdateManagerFactory;
import com.google.android.play.core.install.InstallState;
import com.google.android.play.core.install.InstallStateUpdatedListener;
import com.google.android.play.core.install.model.AppUpdateType;
import com.google.android.play.core.install.model.UpdateAvailability;
import com.google.android.play.core.tasks.Task;

import static android.app.Activity.RESULT_OK;

public class RnInAppUpdateModule extends ReactContextBaseJavaModule implements InstallStateUpdatedListener,
        ActivityEventListener {

    private final ReactApplicationContext reactContext;
    private static final int MY_REQUEST_CODE = 1056;
    private AppUpdateManager appUpdateManager;
    private ReactApplicationContext reactApplicationContext;

    public RnInAppUpdateModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        reactApplicationContext = reactContext;
    }

    @Override
    public String getName() {
        return "RnInAppUpdate";
    }

    @ReactMethod
    private void init() {
        appUpdateManager = AppUpdateManagerFactory.create(reactApplicationContext);
        appUpdateManager.registerListener(this);

        // Returns an intent object that you use to check for an update.
        Task<AppUpdateInfo> appUpdateInfoTask = appUpdateManager.getAppUpdateInfo();


        // and checks the update priority.
        appUpdateInfoTask.addOnSuccessListener(appUpdateInfo -> {
            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE) {
                try {
                    appUpdateManager.startUpdateFlowForResult(
                            appUpdateInfo,
                            AppUpdateType.FLEXIBLE,
                            getCurrentActivity(),
                            MY_REQUEST_CODE);
                } catch (IntentSender.SendIntentException e) {
                    e.printStackTrace();
                    Log.e("mye", "SendIntentException " + e.getMessage());
                }
            }
        });
    }

    @Override
    public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
        if (requestCode == MY_REQUEST_CODE) {
            if (resultCode != RESULT_OK) {
                Log.e("mye", "onActivityResult: " + resultCode);
            }
        }
    }

    @Override
    public void onNewIntent(Intent intent) {

    }

    @Override
    public void onStateUpdate(InstallState state) {

    }
}

File: android/src/main/java/com/rninappupdate/RnInAppUpdateModule.kt
package com.rninappupdate

import android.app.Activity
import android.content.Intent
import android.content.IntentSender
import com.facebook.react.bridge.*
import com.google.android.gms.tasks.Task
import com.google.android.play.core.appupdate.AppUpdateInfo
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability

class RnInAppUpdateModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext), ActivityEventListener {

  private val appUpdateManager: AppUpdateManager = AppUpdateManagerFactory.create(reactContext)

  init {
    reactContext.addActivityEventListener(this)
  }

  override fun getName(): String {
    return NAME
  }

  @ReactMethod
  fun showUpdatePopup(updateType: Int, promise: Promise) {
    val activity: Activity? = currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "Current activity is null")
      return
    }

    val appUpdateInfoTask: Task<AppUpdateInfo> = appUpdateManager.appUpdateInfo

    appUpdateInfoTask.addOnSuccessListener { appUpdateInfo ->
      if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE &&
          appUpdateInfo.isUpdateTypeAllowed(updateType)) {
        try {
          appUpdateManager.startUpdateFlowForResult(
            appUpdateInfo,
            updateType,
            activity,
            REQUEST_CODE
          )
          promise.resolve("STARTED")
        } catch (e: IntentSender.SendIntentException) {
          promise.reject("INTENT_ERROR", "Error starting update flow", e)
        }
      } else {
        promise.reject("UPDATE_NOT_AVAILABLE", "Update not available or type not allowed")
      }
    }

    appUpdateInfoTask.addOnFailureListener { e ->
      promise.reject("UPDATE_ERROR", "Failed to get update info", e)
    }
  }

  override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
    // No-op
  }

  override fun onNewIntent(intent: Intent) {
    // No-op
  }

  companion object {
    const val NAME = "RnInAppUpdate"
    const val REQUEST_CODE = 12345
  }
}
